File: tools/book_mark_format.py
import os
import json
import shutil

def sanitize_filename(filename):
    # Windowsファイル名で許可されていない文字
    invalid_chars = ['<', '>', ':', '"', '/', '\\', '|', '?', '*', ' ']
    
    # 無効な文字を '_' に置換
    for char in invalid_chars:
        filename = filename.replace(char, '_')
    
    return filename

# JSONファイルのパス
home_dir = os.path.expanduser("~")
json_file_path = home_dir + "/AppData/Local/Google/Chrome/User Data/Default/Bookmarks"

try:
    # JSONファイルを開いてデータを読み込む
    with open(json_file_path, 'r', encoding='utf-8') as json_file:
        json_data = json.load(json_file)
    
    # JSONデータを変数に格納しました
    print("JSONデータの読み込みに成功しました:")
    print(json_data)
    
except FileNotFoundError:
    print(f"指定されたJSONファイルが見つかりません: {json_file_path}")
except json.JSONDecodeError as e:
    print(f"JSONデータを解析できません: {e}")
except Exception as e:
    print(f"エラーが発生しました: {e}")

# ブックマークのJSONデータ
bookmark_data = json_data
# ブックマークのルートディレクトリ
root_folder = bookmark_data["roots"]
# ブックマークをNameごとに保存するフォルダーのパス
output_folder = os.path.expanduser("~/bookmark")
# フォルダを作成
if not os.path.exists(output_folder):
    os.mkdir(output_folder)

# ディレクトリが存在することを確認
if os.path.exists(output_folder) and os.path.isdir(output_folder):
    # ディレクトリ内のファイルを列挙し、削除
    for filename in os.listdir(output_folder):
        file_path = os.path.join(output_folder, filename)
        if os.path.isfile(file_path):
            os.remove(file_path)
    
    print(f"{output_folder} 内のファイルを削除しました。")
else:
    print(f"{output_folder} は存在しないか、ディレクトリではありません。")

# ブックマークをNameごとにファイルに保存
for folder_name, folder_data in root_folder.items():
    if "children" in folder_data:
        for bookmark in folder_data["children"]:
            if "name" in bookmark and "url" in bookmark:
                name = bookmark["name"]
                url = bookmark["url"]
                replace_name = sanitize_filename(name)
                with open(os.path.join(output_folder, replace_name), "w") as file:
                    file.write(url)
                
print("===== Done format bookmark json =====")

File: bookmark/book_mark_format.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"os/user"
	"path/filepath"
	"runtime"
	"strings"
	"time"
)

var dirPath string

func main() {
	start := time.Now()

	// ユーザ情報取得
	usr, err := user.Current()
	if err != nil {
		fmt.Println("Error getting user:", err)
		return
	}
	homeDir := usr.HomeDir

	// OSに応じたブックマークJSONのパス設定
	var bookmarkJSON string
	switch runtime.GOOS {
	case "windows":
		bookmarkJSON = "AppData/Local/Google/Chrome/User Data/Default/Bookmarks"
	case "darwin": // Mac OSは"darwin"として検出される
		bookmarkJSON = "Library/Application Support/Google/Chrome/Default/Bookmarks"
	case "linux":
		bookmarkJSON = ".config/google-chrome/Default/Bookmarks"
	default:
		fmt.Println("Unsupported operating system:", runtime.GOOS)
		return
	}

	// ブックマークJSONのパス作成
	JsonPath := filepath.Join(homeDir, bookmarkJSON)

	// JSONの読み込み開始
	file, err := os.Open(JsonPath)
	if err != nil {
		fmt.Println("Error opening JSON file:", err)
		return
	}
	defer file.Close()

	// JSONデータをインターフェースとしてデコード
	var jsonData interface{}
	decoder := json.NewDecoder(file)
	err = decoder.Decode(&jsonData)
	if err != nil {
		fmt.Println("Error decoding JSON:", err)
		return
	}

	// bookmarkディレクトリの存在確認
	dirPath = filepath.Join(homeDir, "bookmark")
	if _, err := os.Stat(dirPath); !os.IsNotExist(err) {
		if err := os.RemoveAll(dirPath); err != nil {
			fmt.Println("Error removing dir:", err)
		}
	}

	// bookmarkディレクトリ作成
	if err := os.Mkdir(dirPath, 0755); err != nil {
		fmt.Println("Error creating dir:", err)
	}

	// ブックマークをNameごとにファイルに保存
	j, ok := jsonData.(map[string]interface{})
	if !ok {
		fmt.Println("Error jsonData type assertion:", err)
		return
	}
	roots, ok := j["roots"]
	if !ok {
		fmt.Println("Error roots type assertion:", err)
		return
	}
	r := roots.(map[string]interface{})
	for _, folderData := range r {
		extractUrls(folderData)
	}

	fmt.Println("===== Done format bookmark json =====")
	end := time.Now()
	total := end.Sub(start)
	fmt.Println("total:", total)
}

func extractUrls(folderData interface{}) {
	f := folderData.(map[string]interface{})
	children, ok := f["children"]
	if !ok {
		fmt.Println("Error children type assertion:")
		return
	}
	c := children.([]interface{})
	for _, folderData := range c {
		f := folderData.(map[string]interface{})

		n, ok := f["name"]
		if !ok {
			continue
		}
		nStr := n.(string)
		name := replaceFilename(nStr)

		u, ok := f["url"]
		if !ok {
			extractUrls(folderData)
			continue
		}
		uStr := u.(string)
		url := []byte(uStr)

		bookmarkFile := filepath.Join(dirPath, name)
		err := os.WriteFile(bookmarkFile, url, 0755)
		if err != nil {
			fmt.Println("Error creating bookmark file:", err)
		}
	}
}

func replaceFilename(name string) string {
	// Windowsのファイル名に使用できない文字
	invalidChars := []string{"<", ">", ":", "\"", "/", "\\", "|", "?", "*", " "}

	// 各文字を "_" に置換
	for _, char := range invalidChars {
		name = strings.ReplaceAll(name, char, "_")
	}

	return name
}
